### ISSUE TITLE
general issue: path data APIs aren't robust to coordinate errors

### ISSUE THREAD

**Issue by eckter** [OpenRailAssociation/osrd#9372]
Most of OSRD exclusively deals with "offsets" to locate elements on their respective tracks or paths. When we want to display them, the front-end is in charge of projecting them on linear geographic elements.

The problem is, we often have mismatches between the referenced track lengths and their linestrings. Imported infrastructures have data quality issues, and generated infrastructures just don't have coordinates that match their distances.

In some cases, we can normalize offsets. For example, an object placed at offset=50m on a track of length=100m, we can place it at the middle of the linestring, even if that linestring is 500m long.

But that doesn't work out well when we need to project data over *a whole train path*. Any data error on any given track will throw off the normalization.

And the issue is that when the front-end projects objects on the path, it mostly relies on the `path_properties` endpoint. That endpoint returns the geographic data for the whole path. Worse, it doesn't even include the expected path length, making it impossible to normalize anything. Most projections end up being wrong.

Causes several bugs:

* https://github.com/OpenRailAssociation/osrd/issues/9228
* https://github.com/OpenRailAssociation/osrd/issues/9110

---

Possible fixes:

We can't think of any *satisfying* fix (but we're open to suggestions). For now we have a few general ideas:

1. Normalize the distances by ranges, included alongside the path geography (e.g. first 200m of geo path is worth 500m of expected path length)
2. All objects returned by the back-end include their geo data, no more front-end projection
3. All offsets include both a "path length" offset and a "geo length" offset

**Comment by maelysLeratRosso** [comment]
Could it also trigger this [bug ](https://github.com/OpenRailAssociation/osrd/issues/10221)?

-> EDIT: Probably not

**Comment by bougue-pe** [comment]
Workshop with @flomonster @clarani @eckter @SharglutDev @maelysLeratRosso @sim51 @Tristramg 

## TLDR

> [!IMPORTANT]  
>
> **This solution is on hold after a possibly better alternative has been expressed (see next comment)**

> Solution on hold (for current bugs and future designs):
> * When backend returns a topological offset/length (ranges) to be projected on a geographical linestring (so far: path):
>    Backend must also provide a corresponding ratio of the considered geographical linestring
> * All inputs to the backend must use only topological length (already OK)
>
> * ⚠ offsets on a tracksection are not impacted by this, as the ratio is pretty straightforward: tracksection contains its length and geographical length (from linestring).

The goals for the workshop are :

    1. make sure this is an actual issue (can/will be handled elsewhere?) -> It is an actual issue
    2. what is impacted by this bug -> /pathfinding for now
    3. build a "generic" target for those cases (based on ROI)
    4. then only after that, determine if/where it should be implemented -> Yes


## 1. is this a real problem we want to address?

For toy infra, we want to be able to simplify the GPS but use topological length.
The front allows that (with a warning).
This is convenient to support those cases.

-> So we should address that at some point


## 2. What is impacted by this bug

List of impacted endpoint (editoast):
1. /pathfinding when manually adding a point, if there is an inconsistency in previous track-section, the point (OP) **was** offset'ed
2. /pathfinding (incompatible constraints errors)
3. ... to be completed

Slightly different problem:
- track-section split: click on a point, then use the GPS length on the track-section, then use proportional topological length to split. There are inconsistencies between `turf` length computation and `postgis`


## 3. Solutions

For outputs:
1. Normalize the distances by ranges, included alongside the path geography (e.g. first 200m of geo path is worth 500m of expected path length): backend publishes a "conversion scale".
2. All objects returned by the back-end include their geo data, no more front-end projection: let the front make most of the computing.
  a. One solution is implemented for the pathfinding in https://github.com/OpenRailAssociation/osrd/pull/10515 (front does extra queries)
3. **All offsets include both a "path length" offset and a "geo length" offset: when backend returns a value (length/offset) it should return both values (to be used when displaying value or GPS)** Use a ratio seems more secure to spot bugs/misusages.

For inputs (we don't want redundant inputs):
1. _not explored_
2. _not explored (or this is mostly solution 4 below)_
3. May also work. Drawback is manipulating multiple types as input.
4. **Always use topological length as input** considering it's (only?) track-sections offset. And reconsider if problem arises.


* for pathfinding:
  * enrich /path_properties for success
  * enrich /pathfinding response with when failing


## 4. determine if/when we should implem this

We want to implem this as bugfixes/enhancement (update/comment the 2 issues already attached to this ticket).
* Starting with #9228 is probably a good idea (implem editoast then replace #10515).
* Doing #9110 only after may help have a quicker idea if moving all enrichment to editoast is better.

The rough evaluation for the work on this (/pathfinding) is around a week.


## Questions leftover (let's see when implementing if they deserve some more thinking)

* Are there only offsets/length on track-sections as backend input ? (when considering a geographical object)
* Should we only implement enrichment of the response in editoast (mutualize, have data available) or just do it "blindly" where it's most convenient? Going for core implem for pathfinding's incompatible constraints.

**Comment by bougue-pe** [comment]
Small workshop held on 2025-03-11, with @flomonster @eckter and @bougue-pe:

A new idea has been expressed (actually refined after a first misunderstanding) :

> [!IMPORTANT]  
>
> ## TLDR
> provide a curve that allows for conversion from topo offset to geom offset (via interpolation) along the path

![Image](https://github.com/user-attachments/assets/39f78595-3082-4a66-8730-0a9fa4c9d834)

Provide a monotonous (and cumulative) curve, to map topologic offset to geometric offset on the path (new attribute in response of `/path_properties` endpoint):
```json
{
  "geom_projection": {
    "topo_offsets": [0, 5000, 12000],
    "geom_offsets": [0, 6000, 9000]
  }
}
```
In this example:
* a first track range is 5m long in topography (`length` attribute) and is 6 m long in geometry (`geo` attribute)
* second track range is 7m long in topo, and 3m long in geometry
* those 2 track ranges might themselves be only part of a track-section (same proportion of track-section is applied for topo and geom)
* a topo offset of 8.5m (middle of second track-range) results in a geom offset of 7.5m

Properties:
  * Same number of values in both list
  * Keep the 0 in both lists for explicitness and ease of use and the last one is the length of the path (respectively topo_length and geom_length). This would also quite match how `path_item_positions` works.
  * Express both values in the same unit as path_item_position: mm (this is actually the target unit for all length in OSRD).

This would be used by any display to be positioned along the path's geometry, the conversion from topo offset being realized by interpolating values along the curve.
Conversion may be done in both directions (no use-case for geom to topo so far).
⚠ Interpolating would require to be resilient to non-continuous cases (track-section with null-length for topo while geom-length would be strictly positive, or the opposite): resulting interpolation in those cases is only required to be in the "valid interval", the center probably being the best choice for the end-user if possible.
It may also be interesting for front to compute the full geom length of the geo shape to confront it to the last value of `geom_offsets` and check if there is an important mismatch (bug? apply ratio then?).

This requires precise documentation (editoast's openapi).

@clarani : Can you check if it is a valid/easy-enough solution for the front? We assumed that curve interpolation is already done in front and shouldn't be a problem, but it requires a validation.

---

This approach is valid for any other geometry.
On a track-section though, computing geom length and obtaining topo length is straight-forward, so no specific attributes are added to description.

Note: We may have the same issue if we want to add back the feature of the train-cursor on the map.

**Comment by clarani** [comment]
All good for me ! ✅

**Comment by flomonster** [comment]
@eckter Is this bug still here, I could not reproduce bug #9228

**Comment by eckter** [comment]
I'd be surprised if this is fixed, considering the API still doesn't allow the front-end to properly handle coordinate error (assuming it didn't change recently).

But I'll look into it. (next week though)

**Comment by bougue-pe** [comment]
> [@eckter](https://github.com/eckter) Is this bug still here, I could not reproduce bug [#9228](https://github.com/OpenRailAssociation/osrd/issues/9228)

#9228 is fixed, but should be refactored, as noted in https://github.com/OpenRailAssociation/osrd/issues/9228#issuecomment-2639494762

We can close #9228 (I kept it open to track refactoring and just changed the type now) if you think it's clearer to have it closed.

The plan after back-end part (https://github.com/OpenRailAssociation/osrd/issues/11191) is:
1. refactor #9228's fix
2. fix #9110
